controllers/machinepool_controller_noderef.go

```diff
@@ -0,0 +1,202 @@
+			"NodeRefs != ReadyReplicas [%q != %q] for MachinePool %q in namespace %q", len(nodeRefsResult.references), machinepool.Status.ReadyReplicas, machinepool.Name, machinepool.Namespace)
+	}
+	return nil
+}
+
+// deleteRetiredNodes deletes nodes that don't have a corresponding ProviderID in Spec.ProviderIDs.
+// A MachinePool infrastucture provider indicates an instance in the set has been deleted by
+// removing its ProviderID from the slice.
+func (r *MachinePoolReconciler) deleteRetiredNodes(ctx context.Context, client client.Client, nodeRefs []apicorev1.ObjectReference, providerIDs []string) error {
+	logger := r.Log.WithValues("providerIDs", len(providerIDs))
+	nodeRefsMap := make(map[string]*apicorev1.Node, len(nodeRefs))
+	for _, nodeRef := range nodeRefs {
+		node := &corev1.Node{}
+		err := client.Get(ctx, types.NamespacedName{Name: nodeRef.Name}, node)
+		if err != nil {
+			return err
+		}
+
+		nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+		if err != nil {
+			logger.Error(err, "Failed to parse ProviderID", "node", node.Name)
+			continue
+		}
+
+		nodeRefsMap[nodeProviderID.ID()] = node
+	}
+	for _, providerID := range providerIDs {
+		pid, err := noderefutil.NewProviderID(providerID)
+		if err != nil {
+			logger.Error(err, "Failed to parse", "providerID", providerID)
+			continue
+		}
+		delete(nodeRefsMap, pid.ID())
+	}
+	for _, node := range nodeRefsMap {
+		err := client.Delete(ctx, node)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, c client.Client, providerIDs []string) (getNodeReferencesResult, error) {
+	logger := r.Log.WithValues("providerIDs", len(providerIDs))
+
+	var ready, available int
+	nodeRefsMap := make(map[string]apicorev1.Node)
+	nodeList := apicorev1.NodeList{}
+	for {
+		err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue))
+		if err != nil {
+			return getNodeReferencesResult{}, err
+		}
+
+		for _, node := range nodeList.Items {
+			nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+			if err != nil {
+				logger.Error(err, "Failed to parse ProviderID", "node", node.Name)
+				continue
+			}
+
+			nodeRefsMap[nodeProviderID.ID()] = node
+		}
+
+		if nodeList.Continue == "" {
+			break
+		}
+	}
+
+	var nodeRefs []apicorev1.ObjectReference
+	for _, providerID := range providerIDs {
+		pid, err := noderefutil.NewProviderID(providerID)
+		if err != nil {
+			logger.Error(err, "Failed to parse", "providerID", providerID)
+			continue
+		}
+		if node, ok := nodeRefsMap[pid.ID()]; ok {
+			available++
+			nodeReady := nodeIsReady(&node)
+			if nodeReady {
+				ready++
+			}
+			nodeRefs = append(nodeRefs, apicorev1.ObjectReference{
+				Kind:       node.Kind,
+				APIVersion: node.APIVersion,
+				Name:       node.Name,
+				UID:        node.UID,
+			})
+		}
+	}
+
+	if len(nodeRefs) == 0 {
+		return getNodeReferencesResult{}, ErrNodeNotFound
```

Comment:
Where is this error defined? I'd have our own error in this file, but with something more specific to MachinePool.

Example: `ErrNoAvailableNodes`